### PR TITLE
Adds ability to import users with phone second factors.

### PR DESCRIPTION
**Pull request by bojeil-google**
WIP: Adds ability to import users with phone second factors.

### CODE REVIEW

File: test/unit/auth/user-import-builder.spec.ts [hunk 1]
```diff
@@ -555,6 +601,77 @@ describe('UserImportBuilder', () => {
...
+        {
+          uid: '1234',
+          multiFactor: {
+            enrolledFactors: [
+              {
+                uid: 'enrolledSecondFactor1',
+                phoneNumber: '+16505557348',
+                displayName: 'Spouse\'s phone number',
+                factorId: 'phone',
+                enrollmentTime: 'invalid',
```
Comment (hiranya911) [inline test/unit/auth/user-import-builder.spec.ts]: I expected this to throw. How is it getting silently ignored?

File: test/unit/auth/user-import-builder.spec.ts [hunk 2]
```diff
@@ -555,6 +601,77 @@ describe('UserImportBuilder', () => {
...
+      const invalidMultiFactorUsers: any = [
+        {
+          uid: '1234',
+          multiFactor: {
+            enrolledFactors: [
+              {
+                uid: 'enrolledSecondFactor1',
+                secret: 'SECRET',
+                displayName: 'Google Authenticator on personal phone',
+                factorId: 'totp',
```
Comment (hiranya911) [inline test/unit/auth/user-import-builder.spec.ts]: Expected to throw. 

It looks like there's some pre-existing logic in the UserImportBuilder to handle these errors, and ignore them. May be we should at least log a warning when this happens, so a simple typo in the user code doesn't go completely unnoticed.

File: src/auth/user-import-builder.ts [hunk 3]
```diff
@@ -193,6 +213,47 @@ function populateUploadAccountUser(
...
   }
+
+  // Convert user.multiFactor.enrolledFactors to server format.
+  if (validator.isNonNullObject(user.multiFactor) &&
+      validator.isNonEmptyArray(user.multiFactor.enrolledFactors)) {
+    user.multiFactor.enrolledFactors.forEach((multiFactorInfo) => {
+      let enrolledAt;
+      if (typeof multiFactorInfo.enrollmentTime !== 'undefined') {
+        if (validator.isUTCDateString(multiFactorInfo.enrollmentTime)) {
+          enrolledAt = new Date(multiFactorInfo.enrollmentTime).toISOString();
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: Added comment.

File: src/auth/auth-api-request.ts [hunk 4]
```diff
@@ -151,6 +151,66 @@ class TenantAwareAuthResourceUrlBuilder extends AuthResourceUrlBuilder {
...
+    }
+  }
+  if (!validator.isNonEmptyString(request.mfaEnrollmentId)) {
+    throw new FirebaseAuthError(
+      AuthClientErrorCode.INVALID_UID,
+      `The second factor "uid" must be a valid non-empty string.`,
+    );
+  }
+  if (typeof request.displayName !== 'undefined' &&
+      typeof request.displayName !== 'string') {
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/auth-api-request.ts]: done

File: src/auth/auth-api-request.ts [hunk 5]
```diff
@@ -381,6 +442,15 @@ function validateCreateEditRequest(request: any, uploadAccountRequest: boolean =
       validateProviderUserInfo(providerUserInfoEntry);
     });
   }
+  // mfaInfo has to be an array of valid AuthFactorInfo requests.
+  if (typeof request.mfaInfo !== 'undefined' &&
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/auth-api-request.ts]: done

File: test/unit/auth/user-import-builder.spec.ts (hunk 1 above)
Comment (bojeil-google) [reply test/unit/auth/user-import-builder.spec.ts]: Actually it only gets ignored when building the request. The client side errors are cached and will be injected in the response alongside server side response. Check test `should return the expected response for import with client side errors`. I added a test for this.

File: test/unit/auth/user-import-builder.spec.ts (hunk 2 above)
Comment (bojeil-google) [reply test/unit/auth/user-import-builder.spec.ts]: Same here. The client side errors are cached and will be injected in the response alongside server side response. Check test `should return the expected response for import with client side errors`. I added a test for this too.

**Comment by bojeil-google** [comment]
Thanks for the quick and helpful review!

**Pull request merged**